Add tests

wKovacs64 opened the pull request:
Closes #1. I'm sure it's not perfect, but it's a start.

```
=============================== Coverage summary ===============================
Statements   : 100% ( 34/34 )
Branches     : 96.43% ( 27/28 )
Functions    : 100% ( 9/9 )
Lines        : 100% ( 34/34 )
================================================================================
```

Thoughts, @tizmagik ?

wKovacs64 committed:
Add initial scaffolding for tests

wKovacs64 committed:
Flip IE8 detection conditionals to ease testing

The 'textContent' property of a jsdom-created style element always seems to be
present (even after attempting to delete), so testing the IE8 case was proving
difficult. Inverting the conditional makes it much easier, as we're now able to
pass in a phony IE8 style element by pre-populating styleSheet.cssText.

wKovacs64 committed:
Add tests for index.js

wKovacs64 committed:
Add tests for insert-style.js

tizmagik commented:
@wKovacs64  Very nice, thank you for this. I'll try and do a full review later today or tomorrow. Could you just add a test for the `@decorator` syntax style please? 

A couple other things to really polish this off (feel free to submit separate issues to do these in follow-on PRs, that's totally fair!)
- Travis build step for PRs
- Coveralls to ensure code coverage doesn't drop
- Pre-publish test to ensure tests pass before npm publish

tizmagik commented:
Please also add yourself as a contributor to the package.json -- you'll get commit access to this repo after this PR is merged. My rule is 2 contributions grants commit access.

wKovacs64 committed:
Add self to contributors

Jeremy made me do it.

wKovacs64 committed:
Include tests in prepublish script

wKovacs64 committed:
Add Travis CI and Coveralls integration

wKovacs64 commented:
Sure, I can add that stuff. I didn't test decorator syntax at first because it felt like we would just be testing Babel's transform, but I'll add one.

wKovacs64 committed:
Add decorator test and rename test components for clarity

tizmagik commented on test/index.test.js:
```diff
@@ -0,0 +1,125 @@
+      const rendered = ReactTestUtils.renderIntoDocument(<WrappedButton />);
+      rendered.componentWillMount();
+      rendered.componentWillUnmount();
+      expect(cache.delete.called).to.false;
```
`to.be.false`?

Why do you expect it to not be called though? Isn't it being unmounted?

wKovacs64 replied:
Good catch, thanks.

It's being unmounted once, but it's been mounted twice (once in the initial `renderIntoDocument` call, and again in the `rendered.componentWillMount` call (manually triggered lifecycle method). So we expect the cache to remain in tact (until the last instance of it has been unmounted). We would have to unmount one more time in order for the cache to clear.

tizmagik replied:
Ah ok got it! That wasn't immediately obvious to me so maybe a comment here would help.

wKovacs64 replied:
Added a comment.

tizmagik commented on test/setup.js:
```diff
@@ -0,0 +1,47 @@
+
+export const testElmIe8 = jsdom.jsdom().createElement('style');
+testElmIe8.setAttribute('type', 'text/css');
+testElmIe8.styleSheet = { cssText: testCss };
```
Maybe this file isn't exactly a "setup" but more of a "TestUtil" or similar might be a better name?

wKovacs64 replied:
Sure. I never know what to call it. :)

tizmagik commented on .travis.yml:
```diff
@@ -0,0 +1,10 @@
+node_js:
+  - "4"
+  - "5"
+  - "node"
```
6?

wKovacs64 replied:
"node" = current, which is 6, but we can future proof it a bit by adding 6 explicitly. I'll add it.

tizmagik commented on src/insert-style.js:
```diff
@@ -1,9 +1,9 @@
 export function getStyle(elm) {
```
Can you help me understand why this file needed to change?

tizmagik replied:
To clarify, I actually like the way you've refactored it, but I just want to understand if this was a necessary change to facilitate the tests or what exactly?

wKovacs64 replied:
Are you referring to flipping the IE8 detection conditionals? I put a note about it in the commit message but I've been rewriting history here like crazy for the PR so it's easy to miss. :) 

The 'textContent' property of a jsdom-created style element always seems to be
present (even after attempting to delete), so testing the IE8 case was proving
difficult. Inverting the conditional makes it much easier, as we're now able to
pass in a phony IE8 style element by pre-populating styleSheet.cssText.

tizmagik replied:
👍

wKovacs64 committed:
Rename test setup module to testUtil

wKovacs64 commented:
Rebased and added/updated the things you mentioned. We might need to mess with some dependencies and add peers (Node 4 is mad), but I'll wait and look into it after this gets merged. I had to add `babel-runtime` to use the decorator transform, but I think it probably should have been included in #11 anyway (if I'm reading [the docs](https://babeljs.io/docs/plugins/transform-runtime) correctly).

Let me know if you want any of these squashed.

tizmagik commented:
This is great, thanks @wKovacs64 !

I wouldn't worry too much about node 4, in fact, you can probably remove it if it doesn't work after merge.
